batch/batch/worker/worker.py

```diff
@@ -1789,6 +1778,259 @@ def __str__(self):
+
+    def __str__(self):
+        return f'JVM-{self.index}'
+
+    def __repr__(self):
+        return f'JVM-{self.index}'
+
+    def interrupt(self):
+        self.should_interrupt.set()
+
+    def reset(self):
+        self.should_interrupt.clear()
+
+    def kill(self):
+        if self.process is not None:
+            self.process.kill()
+
+    def output(self) -> str:
+        return self.process.output()
+
+    def retrieve_and_clear_output(self) -> str:
+        return self.process.retrieve_and_clear_output()
+
+    async def execute(self, classpath: str, mainclass: str, scratch_dir: str, command_string: List[str]):
+        assert worker is not None
+
+        log.info(f'{self}: execute')
+
+        with ExitStack() as stack:
+            reader: asyncio.StreamReader
+            writer: asyncio.StreamWriter
+            reader, writer = await self.new_connection()
+            stack.callback(writer.close)
+            log.info(f'{self}: connection acquired')
+
+            command_string = [classpath, mainclass, scratch_dir, *command_string]
+
+            write_int(writer, len(command_string))
+            for arg in command_string:
+                assert isinstance(arg, str)
+                write_str(writer, arg)
+            await writer.drain()
+
+            wait_for_message_from_process: asyncio.Future = asyncio.ensure_future(read_int(reader))
+            stack.callback(wait_for_message_from_process.cancel)
+            wait_for_interrupt: asyncio.Future = asyncio.ensure_future(self.should_interrupt.wait())
+            stack.callback(wait_for_interrupt.cancel)
+
+            await asyncio.wait([wait_for_message_from_process, wait_for_interrupt], return_when=asyncio.FIRST_COMPLETED)
+
+            if wait_for_interrupt.done():
+                await wait_for_interrupt  # retrieve exceptions
+                if not wait_for_message_from_process.done():
+                    write_int(writer, 0)  # tell process to cancel
+                    await writer.drain()
+
+            eos_exception = None
+            try:
+                message = await wait_for_message_from_process
+            except EndOfStream as exc:
+                try:
+                    self.kill()
+                except ProcessLookupError:
+                    log.warning(f'{self}: JVM died after we received EOS')
+                message = JVM.FINISH_JVM_EOS
+                eos_exception = exc
+
+            if message == JVM.FINISH_NORMAL:
+                log.info(f'{self}: finished normally (interrupted: {wait_for_interrupt.done()})')
+            elif message == JVM.FINISH_CANCELLED:
+                assert wait_for_interrupt.done()
+                log.info(f'{self}: was cancelled')
+            elif message == JVM.FINISH_USER_EXCEPTION:
+                log.info(f'{self}: user exception encountered (interrupted: {wait_for_interrupt.done()})')
+                exception = await read_str(reader)
+                raise ValueError(exception)
```

Comment:
I think this could still log higher up in the callstack, and I don't think we should log user exceptions, both from a noise and security POV. Maybe we should have a `UserError` exception and ensure that we never log exceptions of those types.

Reply:
Agreed, the exception handling is a bit of a mess. It also doesn't get delivered correctly to the user. I've marked a to-do to follow up on this.